ethpm/_utils/cache.py

```diff
@@ -17,8 +17,9 @@ class cached_property:
     def __init__(self, func: Callable[..., Any], name: str = None) -> None:
         self.func = func
         self.name = name or func.__name__
+        self.__doc__ = getattr(func, '__doc__')
         # type ignored b/c self is 'cached_property' but expects 'Callable[..., Any]'
-        functools.update_wrapper(self, func)  # type: ignore
+        functools.update_wrapper(func, self)  # type: ignore
```

Comment:
why were these args swapped?

Reply:
I've updated the PR. Since building the docs generates the following error message:
```sh
sphinx.errors.SphinxWarning: Failed to get a method signature for ethpm.Package.build_dependencies: <ethpm._utils.cache.cached_property object at 0x7f710d9f1b20> is not a callable object
```
I have fixed it to make the object callable by adding the __call__ func.
Because cached_property is a wrapper object, you are right. we need to keep the order of the parameters(wrapper, wrapped).

Reply:
Actually, this class doesn't work at all. It is supposed to convert the returned value of a given function to an instance whose attribute holds that value. I have figured out a solution that works on python >= 3.6 as follows:

```python
from typing import (
    Any,
    Callable,
)

class cached_property:
    """
    Decorator that converts a method with a single self argument into a
    property cached on the instance.

    Optional ``name`` argument allows you to make cached properties of other
    methods. (e.g.  url = cached_property(get_absolute_url, name='url') )
    """

    def __init__(self, func: Callable[..., Any], name: str = None) -> None:
        self.wrapped_func = func
        self.name = name
        self.__doc__ = getattr(func, '__doc__')

    def __get__(self, instance: Any, cls: Any = None) -> Any:
        """
        Call the function and put the return value in instance.__dict__ so that
        subsequent attribute access on the instance returns the cached value
        instead of calling cached_property.__get__().
        """
        if instance is None:
            return self
        res = instance.__dict__[self.name] = self.func(instance)
        return res

    def __set_name__(self, cls: Any = None, name: str = None) -> None:
        """
        The function is called at the time the cls class is created.
        The descriptor would be assigned to name.
        """
        if self.name is None:
            self.name = name
            self.func = self.wrapped_func
        if name != self.name:
            raise TypeError(
                "Unable to assign cached_property for two different names "
                f"(%{self.name} and {name})."
            )
    @staticmethod
    def func():
        raise TypeError(
            'Unable to use an instance of this class without invoking __set_name__!'
        )

import random
class Bar:
    foo = cached_property(lambda _: random.random())

if __name__ == "__main__":
    #print(cached_property.func())
    bar = Bar()
    print(bar.foo)  # 0.6638420043967768
    print(bar.foo)  # 0.6638420043967768
    bar = Bar()
    print(bar.foo)  # 0.5112941702023637
    print(bar.foo)  # 0.5112941702023637
```

Reply:
i will commit these changes

Reply:
Metaprogramming is lit!